flatpak nvidia

failboat78 opened the pull request:
To enable nvidia encoding inside flatpak we need to build ffmpeg with cuda support. This needs the nvcc compiler - which is not easily obtainable via flatpak. The current freedesktop nvidia driver packages have the driver but are missing nvcc. In the future maybe there will be some sdk extension package that can be added to enable this. 

But for now, we can add the hosts cuda folder to the flatpak build env and get it to work. I'm using arch which has a non-standard install location for cuda (/opt/cuda) - other distros might use different. NVCC is also fussy about it's gcc version - 13.2 is the max supported right now. But arch comes with gcc-14 so even without flatpak this requires extra settings. Flatpak actually makes this easier because the runtime used has 13.2 right now. 

So the changes needed are quite small, we just copy cuda into the build folder, and update the pkgconfig to point to it. Then just set some env variables and the build works. Note that NVCC is only needed for the build, not at runtime. 

Improvements that can be made:
Print error if cuda not installed - or continue without nvidia support
Detect where host has cuda installed, update paths to reflect

Testing:
Works well for me with arch+flatpak steam+4090 555+kde plasma 6.1.
Does this also work for amd users? Would there be an nvidia vs non-nvidia version needed?

failboat78 commented:
Oh yeah I could not get the desktop shortcut part to work. The copy command just says folder does not exist - not sure how to do this properly so I just commented it out.

Meister1593 commented:
Honestly i thought about changing flatpak to just instructions on how to run portable under flatpak container, since it seemed to work for nvidia users

Meister1593 commented:
What formatter was used for json?

failboat78 commented:
Just vscode defaults, sorry for mangling it!

And yes - this flatpak should work for all users - but I have no AMD to test on. This is same method blender and other programs support nvidia via flatpak.

Meister1593 commented on alvr/xtask/flatpak/com.valvesoftware.Steam.Utility.alvr.json:
```diff
@@ -1,64 +1,57 @@
+        "cargo xtask prepare-deps --platform linux",
+        "cargo xtask build-streamer --gpl --release",
+        "cp -r /run/build/alvr/build/alvr_streamer_linux/* /app/utils/alvr",
+        "echo this line does not work - cp -r /run/build/alvr/alvr/xtask/flatpak/com.valvesoftware.Steam.Utility.alvr.desktop ~/.local/share/applications/com.valvesoftware.Steam.Utility.alvr.desktop"
```
tbh we probably gonna make this work through launcher later on

failboat78 commented:
Another improvement is to download cuda rather than copy from host: https://developer.nvidia.com/cuda-toolkit-archive

This would be quite elegant then, and provide a completely portable build environment?

zarik5 commented:
@failboat78 Unfortunately nvidia requires you to give personal details before download. So we cannot automate it. From experience, installing cuda packages may break the system, when not provided by the distro package manager. So maybe we should add the option to link from a user provide directory?

failboat78 commented:
> @failboat78 Unfortunately nvidia requires you to give personal details before download. So we cannot automate it. From experience, installing cuda packages may break the system, when not provided by the distro package manager. So maybe we should add the option to link from a user provide directory?

They must have changed their website, it's possible to download those cuda archives without a login. It does want you to accept license when you install tho, but this can be automated.

Test: wget https://developer.download.nvidia.com/compute/cuda/12.4.1/local_installers/cuda_12.4.1_550.54.15_linux.run

Using flatpak we avoid the usual "breaking the system" problems. We are just using this cuda to do our build, then discarding it. There should be no system to break. 

I'll commit improvements in near future.

failboat78 commented:
@zarik5  @Meister1593  now it should be fully automated portable build. 

But cuda download is 4.1GB so can cause large delay dependent on internet speed.

I was not able to run the cuda installer in the normal fashion - it would segfault. As a workaround I had to use the raw extract option which is provided for weird cases like this. The nvidia software is split between several folders internally - so I had to update pkgconfig to use multiple folders. 

Pretty tedious stuff!

Meister1593 commented:
Same flatpak file is also used for pipelines, i have a feeling this needs to be set up with caching for sure, and maybe even separated?...

failboat78 commented:
Yes we need to cache this cuda stuff. I am trying to figure out how best to do that with flatpak - maybe moving cuda to it's own extension, or another module.

The-personified-devil commented on alvr/server/cpp/platform/linux/EncodePipelineVAAPI.cpp:
```diff
@@ -171,6 +171,8 @@ alvr::EncodePipelineVAAPI::EncodePipelineVAAPI(
         throw std::runtime_error("failed to allocate VAAPI encoder");
     }
 
+    encoder_ctx->gop_size = INT_MAX;
```
This should not be in this PR

failboat78 replied:
I rebased and dropped that commit

Meister1593 reviewed (changes requested):
I'm overall fine with this, but i still wonder if you can just run launcher + alvr under flatpak without any of this
Could you try to:
1. Downloading launcher https://github.com/alvr-org/ALVR/releases/download/v20.9.1/alvr_launcher_linux.tar.gz
2. Putting it into `~/.var/app/com.valvesoftware.Steam`
3. Launching it with `flatpak run --command='~/.var/app/com.valvesoftware.Steam/ALVR Launcher' com.valvesoftware.Steam` (or similar to just be able to run it under runtime)
4. Installing alvr stable version, trying to run it under flatpak steam all as usual with all fixes from troubleshooting

failboat78 commented:
Sure - I will start looking at how to include the launcher with flatpak. 

But that seems like a seperate task - not related to all the nvidia nonsense here. Can we please merge this one and I will open a new PR for launcher flatpak stuff?

Meister1593 commented:
> @Meister1593
> 
> Sure - I will start looking at how to include the launcher with flatpak.
> 
> But that seems like a seperate task - not related to all the nvidia nonsense here. Can we please merge this one and I will open a new PR for launcher flatpak stuff?
> 
> If we don't use the stuff here (which is purely build related) then that is fine - I just wanted to learn more about the build process. And the same stuff can be used to package the launcher too hopefully. What I am clueless about is the interaction between steam vs steamvr vs alvr vs launcher - this requires some study.

Then only one question, related to cache, on @zarik5 - is it fine? It weights quite a bit as far as i can tell and there was an issue with space we had available.

If it's impossible, then launcher could be the only option

zarik5 commented:
If cache is a problem we can remove the MSRV jobs

Meister1593 commented:
> I'm overall fine with this, but i still wonder if you can just run launcher + alvr under flatpak without any of this Could you try to:
> 
>     1. Downloading launcher https://github.com/alvr-org/ALVR/releases/download/v20.9.1/alvr_launcher_linux.tar.gz
> 
>     2. Putting it into `~/.var/app/com.valvesoftware.Steam`
> 
>     3. Launching it with `flatpak run --command='~/.var/app/com.valvesoftware.Steam/ALVR Launcher' com.valvesoftware.Steam` (or similar to just be able to run it under runtime)
> 
>     4. Installing alvr stable version, trying to run it under flatpak steam all as usual with all fixes from troubleshooting

Just double checked that case on amd - it works completely fine

Meister1593 commented:
> If cache is a problem we can remove the MSRV jobs

don't think removing them is a good idea?... they are pretty useful

zarik5 commented:
> don't think removing them is a good idea?... they are pretty useful

We can make so it runs only for release branches or remove its (rust) cache.

Meister1593 commented:
looking at last commit... launcher works with nvidia fine?

Also i wondered if this works we can actually build launcher inside flatpak and push it into flathub?

failboat78 commented:
@Meister1593 Ok I added the launcher to it. And it seems to work as expected. 

Amusingly that cuda build stuff was totally broken, it was copying some unstaged local files I had. So I fixed it - but it seems we don't care and most of this will be discarded. Only the launcher binary goes in the flatpak.

I updated the .desktop file to have the correct run command - but I'm not sure how we can automate the adding of this file to the users desktop environment. Flatpak must offer a way to handle this detail - I'll keep reading. 

Build command: flatpak run org.flatpak.Builder --user --install --force-clean .flatpak-build-dir alvr/xtask/flatpak/com.valvesoftware.Steam.Utility.alvr.json
Run command: flatpak run --command=/app/utils/alvr/bin/alvr_launcher com.valvesoftware.Steam

failboat78 commented:
To add the .desktop file we would copy it to /app/share/applications during the build.

But because we are using the steam runtime, this folder is mounted as read-only. You can see the steam.desktop file in that folder. 

Possible solutions:  
If the steam flatpak marks this folder as an extension point, we can then use "merge-dirs" to add our file to it. 
User manually adds desktop file 

https://docs.flatpak.org/en/latest/conventions.html

Meister1593 commented:
I think we can push it as is without auto-installing desktop file
This is anyway better than before, with nvidia working

Just update flatpak wiki page and we can merge it

Meister1593 reviewed (changes requested):
Revert removal of flatpak page (it's useful), make it just specific for launcher and note how you can install desktop file

zarik5 commented:
@failboat78 ping?

failboat78 commented:
I'm confused - what changes are we doing? Do we want a completely seperate PR for the launcher stuff? This original PR was mostly concerned with using flatpak as a build environment - which is very different. 

If that's what we want I will revert and make a new PR for the launcher.

Meister1593 reviewed (commented):
I think the only thing left is to update the wiki and that's it

@The-personified-devil should it be in the same pr or could be done separately? (just to not stall pr), what do you think?

The-personified-devil commented:
> @The-personified-devil should it be in the same pr or could be done separately? (just to not stall pr), what do you think?

merge this first, wiki prs are really quick to merge

Meister1593 commented:
Adb working under flatpak probably requires dashboard-side changes (to run different command), for now i think it can be skipped, just note that in wiki/readme page later
https://github.com/skylot/jadx/issues/2214 flatpak-spawn might be useful there

failboat78 commented:
I got adb working under flatpak and updated readme. 

However I am still not able to run a complete VR test because it seems a lot of stuff graphics and steamvr got updated in last few weeks.

Meister1593 commented on alvr/xtask/flatpak/README.md:
````diff
@@ -21,20 +19,62 @@ git clone https://github.com/alvr-org/ALVR.git
 ```
-flatpak run org.flatpak.Builder --user --install --force-clean .flatpak-build-dir alvr/xtask/flatpak/com.valvesoftware.Steam.Utility.alvr.json
+cd alvr/xtask/flatpak
+./build_and_install.sh  
````
this script requires host flatpak builder, but before it was from flatpak itself

Meister1593 replied:
`flatpak-builder` vs `flatpak run org.flatpak.Builder`

Meister1593 replied:
Besides this, everything else works fine on amdgpu
If this changed it can be merged just fine

Meister1593 reviewed (approved):
Will merge for now, update nightly and check if it builds correctly there as well (you can try building it on your repo (ignoring android build) and see if flatpak is correct)
If it works, then most likely release build should work fine as well.

And of course, update wiki too.
